File: lcof/cc/xuan-zhuan-shu-zu-de-zui-xiao-shu-zi-lcof.cc
// https://leetcode-cn.com/problems/xuan-zhuan-shu-zu-de-zui-xiao-shu-zi-lcof/

#include <iostream>
#include <vector>

using namespace std;

class Solution
{
public:
    int minArray(vector<int> &numbers)
    {
        size_t left = 0, mid = 0, right = numbers.size() - 1;
        while (left < right)
        {
            mid = (left + right) >> 1;
            if (numbers[mid] > numbers[right])
                left = mid + 1;
            else if (numbers[left] == numbers[right])
                right--;
            else
                right = mid;
        }
        return numbers[left];
    }
};

int main(int argc, char const *argv[])
{
    Solution s;
    vector<int> nums = {2, 2, 2, 0, 1};
    cout << s.minArray(nums) << endl;
    return 0;
}

File: lcof/rs/xuan-zhuan-shu-zu-de-zui-xiao-shu-zi-lcof.rs
// https://leetcode-cn.com/problems/xuan-zhuan-shu-zu-de-zui-xiao-shu-zi-lcof/

#[derive(Debug)]
struct Solution {
}

impl Solution {
    pub fn min_array(numbers: Vec<i32>) -> i32 {
        let mut left = 0usize;        
        let mut right = numbers.len() - 1usize;

        while left < right {
            let mid = (left + right) >> 1;
            if numbers[mid] > numbers[right] {
                left = mid + 1;
            } else if numbers[mid] == numbers[right] {
                right -= 1;
            } else {
                right = mid;
            }
        }

        numbers[left]
    }
}

fn main() {
    let nums = vec![3, 4, 5, 1, 2];
    println!("{:?}", Solution::min_array(nums));
}